drivers/DeviceKey.cpp

```diff
@@ -0,0 +1,255 @@
+{
+    return write_key_to_nvstore(value, isize);
+}
+
+int DeviceKey::write_key_to_nvstore(uint32_t *input, size_t isize)
+{
+    if (DEVICE_KEY_16BYTE != isize && DEVICE_KEY_32BYTE != isize) {
+        return DEVICEKEY_INVALID_KEY_SIZE;
+    }
+
+    //First we read if key exist. If it is exists, we return DEVICEKEY_ALREADY_EXIST error
+    uint32_t read_key[DEVICE_KEY_32BYTE / sizeof(uint32_t)] = {0};
+    size_t read_size = DEVICE_KEY_32BYTE;
+    int ret = read_key_from_nvstore(read_key, read_size);
+    if (DEVICEKEY_SUCCESS == ret) {
+        return DEVICEKEY_ALREADY_EXIST;
+    }
+    if (DEVICEKEY_NOT_FOUND != ret) {
+        return ret;
+    }
+
+    NVStore& nvstore = NVStore::get_instance();
+    ret = nvstore.set(NVSTORE_DEVICEKEY_KEY, (uint16_t)isize, input);
+    if (NVSTORE_WRITE_ERROR == ret || NVSTORE_BUFF_TOO_SMALL == ret) {
+        return DEVICEKEY_SAVE_FAILED;
+    }
+
+    if (NVSTORE_SUCCESS != ret) {
+        return DEVICEKEY_NVSTORE_UNPREDICTABLE_ERROR;
+    }
+
+    return DEVICEKEY_SUCCESS;
+}
+
+int DeviceKey::read_key_from_nvstore(uint32_t *output, size_t& size)
+{
+    if (size > UINT16_MAX) {
+        return DEVICEKEY_INVALID_PARAM;
+    }
+
+    uint16_t in_size = size;
+    uint16_t out_size = 0;
+    NVStore& nvstore = NVStore::get_instance();
+    int nvStatus = nvstore.get(NVSTORE_DEVICEKEY_KEY, in_size, output, out_size);
+    if (NVSTORE_NOT_FOUND == nvStatus) {
+        return DEVICEKEY_NOT_FOUND;
+    }
+
+    if (NVSTORE_READ_ERROR == nvStatus || NVSTORE_BUFF_TOO_SMALL == nvStatus) {
+        return DEVICEKEY_READ_FAILED;
+    }
+
+    if (NVSTORE_SUCCESS != nvStatus) {
+        return DEVICEKEY_NVSTORE_UNPREDICTABLE_ERROR;
+    }
+
+    size = out_size;
+    return DEVICEKEY_SUCCESS;
+}
+
+// Calculate CMAC functions - wrapper for mbedtls start/update and finish
+int DeviceKey::calculate_cmac(const unsigned char *input, size_t isize, uint32_t *ikey_buff, int ikey_size,
+                              unsigned char *output)
+{
+    int ret;
+    mbedtls_cipher_context_t ctx;
+
+    mbedtls_cipher_type_t mbedtls_cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
+    if (DEVICE_KEY_32BYTE == ikey_size) {
+        mbedtls_cipher_type = MBEDTLS_CIPHER_AES_256_ECB;
+    }
+
+    const mbedtls_cipher_info_t *cipher_info = mbedtls_cipher_info_from_type(mbedtls_cipher_type);
+
+    mbedtls_cipher_init(&ctx);
+    ret = mbedtls_cipher_setup(&ctx, cipher_info);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_starts(&ctx, (unsigned char *)ikey_buff, ikey_size * 8);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_update(&ctx, input, isize);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    ret = mbedtls_cipher_cmac_finish(&ctx, output);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    return DEVICEKEY_SUCCESS;
+
+finish:
+    mbedtls_cipher_free( &ctx );
+    return ret;
```

Comment:
You are returning a mixture of own error codes and Mbed TLS specific ones here -- is this intended?